components/patient-data/migrations/src/main/java/org/phenotips/data/internal/R71499PhenoTips2777DataMigration.java

```diff
@@ -0,0 +1,250 @@
+    }
+
+    /**
+     * For each {@code genesymbol} field, rename it to {@code gene} and swtich the value from HGNC symbol to Ensembl ID,
+     * if provided.
+     *
+     * @param doc XWiki document
+     * @param geneVariantClassReference reference to {@code GeneVariantClass}
+     * @throws XWikiException if property value cannot be set
+     */
+    private void migrateGeneVariants(final XWikiDocument doc, final DocumentReference geneVariantClassReference)
+        throws XWikiException
+    {
+        final List<BaseObject> variants = doc.getXObjects(geneVariantClassReference);
+        if (variants == null) {
+            return;
+        }
+        for (final BaseObject variant : variants) {
+            final StringProperty oldGeneName = setPropertyValue(variant, OLD_GENE_NAME);
+
+            // Rename all "genesymbol" properties to "gene".
+            if (oldGeneName != null) {
+                variant.removeField(OLD_GENE_NAME);
+                final StringProperty newGeneName = (StringProperty) oldGeneName.clone();
+                newGeneName.setName(GENE_NAME);
+                variant.addField(GENE_NAME, newGeneName);
+            }
+        }
+    }
+
+    /**
+     * Sets the value to Ensembl ID for a given property.
+     *
+     * @param baseObject {@code BaseObject} containing the property name
+     * @param propertyName the name of the property, for example {@code gene}
+     * @return {@code StringProperty} for the provided property name
+     * @throws XWikiException if property value cannot be set
+     */
+    private StringProperty setPropertyValue(final BaseObject baseObject, final String propertyName)
+        throws XWikiException
+    {
+        if (baseObject == null) {
+            return null;
+        }
+
+        final StringProperty oldBaseObjProp = (StringProperty) baseObject.get(propertyName);
+        // Want to return oldBaseObjProp in case it is not null, that way it can still be renamed if the
+        // genesymbol property, for whatever reason, contains no value.
+        final String geneSymbol = oldBaseObjProp.getValue();
+        if (StringUtils.isBlank(geneSymbol)) {
+            return oldBaseObjProp;
+        }
+
+        final String ensemblId = getEnsemblId(geneSymbol);
+        baseObject.setStringValue(propertyName, ensemblId);
+        return oldBaseObjProp;
+    }
+
+    /**
+     * Gets EnsemblID corresponding to the HGNC symbol.
+     * @param geneSymbol the string representation of a gene symbol (e.g. NOD2).
+     * @return the string representation of the corresponding Ensembl ID.
+     */
+    private String getEnsemblId(final String geneSymbol) {
+        final VocabularyTerm term = this.hgnc.getTerm(geneSymbol);
+        @SuppressWarnings("unchecked")
+        final List<String> ensemblIdList = term != null ? (List<String>) term.get("ensembl_gene_id") : null;
+        final String ensemblId = ensemblIdList != null && !ensemblIdList.isEmpty() ? ensemblIdList.get(0) : null;
```

Comment:
is get(0) always guaranteed to be the right term? E.g. is the first element of the results array always the correct choice? What if there is NO exact match, and the first is just the least inaccurate match? We may have free-text genes (right?) which have no exact match but do have some partial match

E.g. http://localhost:8080/get/PhenoTips/GeneNameService?outputSyntax=plain&rand=0.38996313003383465&json=true&q=GS

returns stuff, but none of it is the correct ensemble ID

i think the correct thing to do is to compare the `symbol`, if the match is exact use the ID, otherwise keep free text as is